### PR TITLE
Build context as part of GenericRun

### LINKED ISSUE #6186: Build the Context as part of the GenericRun
/kind feature

## Which functionality do you think we should add?

Get the context as part of the `genericclioptions.GenericRun` function, instead of calling `o.Context, err = genericclioptions.New(params)` in every command code, and pass the context as part of the ctx received by Run (and make Complete and Validate also receive this context).

Related to #6185 

Context:
- application: "app"
- component name
- directory
- devfile name (.../devfile.yaml or .../.devfile.yaml)
- devfile object
- output flag

### PR DESCRIPTION
**Pull request by feloy**
**What type of PR is this:**

/kind cleanup

**What does this PR do / why we need it:**

**Which issue(s) this PR fixes:**

Fixes #6186 , #5517

**PR acceptance criteria:**

- [ ] Unit test 

- [ ] Integration test 

- [ ] Documentation 

**How to test changes / Special notes to the reviewer:**

### CODE REVIEW

File: pkg/odo/cli/dev/dev.go [hunk 1]
```diff
@@ -214,11 +168,11 @@ func (o *DevOptions) HandleSignal() error {
...
 
-func (o *DevOptions) Cleanup(commandError error) {
+func (o *DevOptions) Cleanup(ctx context.Context, commandError error) {
 	if commandError != nil {
-		devFileObj := o.Context.EnvSpecificInfo.GetDevfileObj()
-		componentName := o.GetComponentName()
-		_ = o.clientset.WatchClient.CleanupDevResources(devFileObj, componentName, log.GetStdout())
+		devFileObj := odocontext.GetDevfileObj(ctx)
+		componentName := odocontext.GetComponentName(ctx)
+		_ = o.clientset.WatchClient.CleanupDevResources(ctx, *devFileObj, componentName, log.GetStdout())
```
Comment (rm3l) [inline pkg/odo/cli/dev/dev.go:175]: Would it make sense to change the `watch` interface and implementation, so clients could just pass a Context to `CleanupDevResources`, e.g.:
```go
_ = o.clientset.WatchClient.CleanupDevResources(ctx, log.GetStdout())
```

**Review by rm3l — CHANGES_REQUESTED** [review]
Added a few comments I had so far. I am yet to review the other commits ;-)

File: pkg/odo/cli/dev/dev.go (hunk 1 above)
Comment (feloy) [reply pkg/odo/cli/dev/dev.go:175]: There are many places where this kind of refactoring needs to be done, I would prefer to do this in specific PRs, to not add more to this one.

We have to take care in which packages we want to propagate this context (for example, the odocontext.Get* functions should not be used in low-level packages like kclient, exec, sync, etc but only on odo-specific packages.

Comment (rm3l) [reply pkg/odo/cli/dev/dev.go:175]: Makes sense - thanks.

**Review by rm3l — APPROVED** [review]
LGTM - it was quite interesting to follow the narrative of this quite large PR by reviewing it commit-by-commit. It didn't require too much context-switch. Thanks.

**Comment by dharmit** [comment]
/override ci/prow/v4.11-integration-e2e 

because the tests are passing on IBM Cloud.

**Pull request merged**